lib/LanguageServerRename/Model/Renamer.php

```diff
@@ -0,0 +1,314 @@
+use Phpactor\ReferenceFinder\Exception\CouldNotLocateDefinition;
+use Phpactor\ReferenceFinder\ReferenceFinder;
+use Phpactor\TextDocument\ByteOffset;
+use Phpactor\TextDocument\Location;
+use Phpactor\TextDocument\TextDocument;
+use Phpactor\TextDocument\TextDocumentBuilder;
+use function mb_strlen;
+use function mb_substr;
+
+class Renamer
+{
+    /**
+     * @var Parser
+     */
+    private $parser;
+    /**
+     * @var DefinitionLocator
+     */
+    private $definitionLocator;
+    /**
+     * @var ReferenceFinder
+     */
+    private $finder;
+    /**
+     * @var ClientApi
+     */
+    private $clientApi;
+    /**
+     * @var int
+     */
+    private $timeoutSeconds = 10;
+    /**
+     * @var Workspace
+     */
+    private $workspace;
+    /**
+     * @var RenameVariable
+     */
+    private $renameVariable;
+    /**
+     * @var NodeUtils
+     */
+    private $nodeUtils;
+
+    public function __construct(
+        Workspace $workspace,
+        Parser $parser,
+        ReferenceFinder $finder,
+        DefinitionLocator $definitionLocator,
+        ClientApi $clientApi,
+        RenameVariable $renameVariable,
+        NodeUtils $nodeUtils
+    ) {
+        $this->parser = $parser;
+        $this->definitionLocator = $definitionLocator;
+        $this->finder = $finder;
+        $this->clientApi = $clientApi;
+        $this->workspace = $workspace;
+        $this->renameVariable = $renameVariable;
+        $this->nodeUtils = $nodeUtils;
+    }
+
+    public function prepareRename(TextDocumentItem $textDocument, Position $position): ?Range
+    {
+        [$offset, $node] = $this->documentAndPositionToNodeAndOffset($textDocument, $position);
+        
+        if ($this->canRenameNode($node)) {
+            return $this->nodeUtils->getNodeNameRange($node);
+        }
+
+        return null;
+    }
+
+    public function rename(TextDocumentItem $textDocument, Position $position, string $newName): ?WorkspaceEdit
+    {
+        [$offset, $node] = $this->documentAndPositionToNodeAndOffset($textDocument, $position);
+        
+        $phpactorDocument = TextDocumentBuilder::create($textDocument->text)
+            ->uri($textDocument->uri)
+            ->language($textDocument->languageId ?? 'php')
+            ->build();
+
+        if ($this->canRenameNode($node))
+            return $this->renameNode($phpactorDocument, $offset, $node, $this->nodeUtils->getNodeNameText($node, $textDocument->text), $newName);
+        
+        return null;
+    }
+    
+    private function renameNode(TextDocument $phpactorDocument, ByteOffset $offset, Node $node, string $oldName, string $newName): ?WorkspaceEdit
+    {
+        if (empty($oldName)) {
+            return null;
+        }
+        $locations = [];
+        try {
+            $potentialLocation = $this->definitionLocator->locateDefinition($phpactorDocument, $offset);
+            $locations[] = new Location($potentialLocation->uri(), $potentialLocation->offset());
```

Comment:
the finder should also find this location, why do we do it here? (probably not a good idea to use the definition locator here, which may do more than we want and shouldn't be coupled to this service)

Reply:
I agree, but the `ReferenceFinder` doesn't return the definition. I took this code from the "Find all references" hander.